sympy/functions/elementary/piecewise.py

```diff
@@ -195,242 +310,422 @@ def _eval_derivative(self, x):
-           update the new conds interval.
-           -  eg x < 1, x < 3 -> [oo,1],[1,3] instead of [oo,1],[oo,3]
-        3) Sort the intervals to make it easier to find correct exprs
-
-        Under normal use, we return the expr,cond pairs in increasing order
-        along the real axis corresponding to the symbol sym.  If targetcond
-        is given, we return a list of (lowerbound, upperbound) pairs for
-        this condition."""
+                neg = -self._eval_interval(x, hi, lo, _first=False)
+                if (lo < hi) is S.false or lo is S.Infinity or hi is S.NegativeInfinity:
+                    rv = neg
+                else:
+                    rv = Piecewise(
+                        (self._eval_interval(x, lo, hi, _first=False),
+                            lo < hi),
+                        (neg,
+                            True))
+                if rv == Undefined:
+                    raise ValueError("Can't integrate across undefined region.")
+                return rv
+
+        # solve a Piecewise with lo <= hi and no x-independent relationals
+        # ------------------------------------------------------------------
+        try:
+            abei = self._abei(x)
+        except NotImplementedError:
+            from sympy import Integral
+            return Integral(self, (x, lo, hi))  # unevaluated
+
+        pieces = [(a, b) for a, b, _, _ in abei]
+        done = [(lo, hi, -1)]
+        oo = S.Infinity
+        for k, p in enumerate(pieces):
+            if p[:2] == (-oo, oo):
+                # all undone intervals will get this key
+                for j, (a, b, i) in enumerate(done):
+                    if i == -1:
+                        done[j] = a, b, k
+                break  # nothing else to consider
+            N = len(done) - 1
+            for j, (a, b, i) in enumerate(reversed(done)):
+                if i == -1:
+                    j = N - j
+                    done[j: j + 1] = _clip(p, (a, b), k)
+        # check for holes
+        for a, b, i in done:
+            if i == -1:
+                if a != b:
+                    # this requires a default (i == -1)
+                    # over a potentially non-zero interval
+                    # since a != b so we can't do this
+                    return S.NaN
+                else:
+                    # XXX are there expressions that would give
+                    # a value over a zero-width interval?
+                    # Integrating DiracDelta or Heaviside
+                    # does but we are not integrating right
+                    # now. If there are things to check for
+                    # check expression given by abei[i][-2]
+                    pass
+        # return the sum of the intervals
+        sum = S.Zero
+        for a, b, i in done:
+            sum += abei[i][-2]._eval_interval(x, a, b)
+        return _mmsimp(sum)
+
+    def _abei(self, sym):
```

Comment:
yes...I've still got a nagging todo to return intervals that are open or closed instead of simply a and b.